### PR TITLE
Fix subquery "Item with Same Key has already been added” exception for Linq provider

### LINKED ISSUE #2659: IQueryable<T> filter by subquery gives "Item with Same Key has already been added”
Execute this query, gives the following exception:
```
statelessSession.Query<Sale>()
    .Select(x => new SaleDto
    {
        Id = x.Id,
        SaleDate = x.ListChildren.Max(y => y.Date)
    })
    .Where(x => x.SaleDate <= DateTime.Today && x.SaleDate >= DateTime.Today)
    .ToList();
```
Exception:
`System.ArgumentException: Item with Same Key has already been added.`

### PR DESCRIPTION
**Pull request by maca88**
Fixes #2659

### CODE REVIEW

File: src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs [hunk 1]
```diff
@@ -317,8 +317,8 @@ protected HqlTreeNode VisitBinaryExpression(BinaryExpression expression)
 			var rightType = GetExpressionType(expression.Right);
 			if (leftType != null && leftType == rightType)
 			{
-				_notCastableExpressions.Add(expression.Left, leftType);
-				_notCastableExpressions.Add(expression.Right, rightType);
+				_notCastableExpressions.TryAdd(expression.Left, leftType);
```
Comment (hazzik) [inline src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs:320]: TryAdd is not available in .net standard 2.0 and .net 4.6.1

Comment (maca88) [reply src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs:320]: Added an extension method for them.

**Review by fredericDelaporte — APPROVED** [review]

File: src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs (hunk 1 above)
Comment (hazzik) [reply src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs:320]: @maca88 can we replace Add with Set? 

_notCastableExpressions[expression.Right] = rightType;

**Comment by hazzik** [comment]
@maca88 one last question from me, see in the comments section.

File: src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs (hunk 1 above)
Comment (maca88) [reply src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs]: Somehow I overlooked the comment. Yes we can, replaced with Set. 
Also now I realized that I created the branch to this repository instead on my fork. We should remove the branch after merging.

**Review by hazzik — APPROVED** [review]

File: src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs (hunk 1 above)
Comment (hazzik) [reply src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs]: >Somehow I overlooked the comment.

This is what I thought.

**Pull request merged**